### DISCUSSION TITLE
Using fftdeconvol

### DISCUSSION THREAD

**Discussion by andrixnet** [ImageMagick/ImageMagick#6780] [Digital Image Processing] [answered]
I'm trying to apply fftdeconvol script to some real workd blurred images I have. 
Mine originates from camera. 
The instructions and examples at http://www.fmwconcepts.com/imagemagick/fftdeconvol/index.php work fine. 
Running fftdeconvol on the sample images shown there works fine.
Running it on my image results in a meaningless pattern of noise.

As per example, I ran ` fftdeconvol infile filtfile outfile` using the cameraman example files and the filter file generated with `convert -size 128x128 xc:black -fill white -draw "circle 64,64 67,64" -alpha off defocus7.png`. It works like charm. 

Now my image is a 4000x3000. First I cut it down to a square 3000x3000 (having previously read that non-square images and fft in ImageMagick yield problems.
Then I generate a filter like this: `convert -size 3000x3000 xc:black -fill white -draw "circle 1500,1500 1515,1500" -alpha off defocus.png`

then I run `fftdeconvol myimg.jpg defocus.png myimg_deblur.png`

I have ImageMagick 6.8.6-10 with HDRI and fftw delegate and quantum_depth=32.

I also tried the cameradeblur which works on the samples from Fred but not on my image.
![img_2510sq_deblur](https://github.com/ImageMagick/ImageMagick/assets/3382777/96ae007c-402d-4c10-8455-c8c5876205f2)

My math understanding is limited, obviously the intuitive application of these scripts seem to fail. What am I doing wrong and what can I make to get proper results? Thank you.

**Comment by fmw42** [comment]
Please provide your input image.  Note that you need a filtfile that accurate represents the degradation in your image. If you have some unknown blur, then you need to use the cepstrum script on your input to figure out the needed degradation and arguments for creating the filtfile.

**Comment by andrixnet** [comment]
Testing with the examples from http://www.fmwconcepts.com/imagemagick/fftdeconvol/index.php

**cameraman2_defocus7.png**
![cameraman2_defocus7](https://github.com/ImageMagick/ImageMagick/assets/3382777/d4781567-370e-419b-8acc-b8db7c17213f)

Filter file generated with: `convert -size 128x128 xc:black -fill white -draw "circle 64,64 67,64" -alpha off defocus7.png`
**defocus7.png**
![defocus7](https://github.com/ImageMagick/ImageMagick/assets/3382777/e55692d6-aded-4a74-8264-b2a973d368a3)

`fftdeconvol cameraman2_defocus7.png defocus7.png cameraman2.png`

The result: 
**cameraman2.png**
![cameraman2](https://github.com/ImageMagick/ImageMagick/assets/3382777/8cfe528d-8dae-4c45-b435-5647f3432a29)

So this works fine.

**Comment by andrixnet** [comment]
Real works image, cropped to square: 
**img_2510sq.jpg**
![img_2510sq](https://github.com/ImageMagick/ImageMagick/assets/3382777/7160f6be-43a5-4fc1-8039-5b0711fa5ba0)

filter file create with: `convert -size 3000x3000 xc:black -fill white -draw "circle 1500,1500 1515,1500" -alpha off defocus.png`
**defocus.png**
![defocus](https://github.com/ImageMagick/ImageMagick/assets/3382777/e3cfddd4-c908-4e64-ba6d-d2f1202b17b0)

fftdeconvol img_2510sq.jpg defocus.png img_2510sq_deblur.jpg

**img_2510sq_deblur.jpg**
![img_2510sq_deblur](https://github.com/ImageMagick/ImageMagick/assets/3382777/e7063828-c307-4e51-b443-b344b674e9e6)

(using PNG as output can't be uploaded here because it's 16bpp)

**Comment by fmw42** [comment]
You cannot just pick any circle.  If you have some unknown blur, then you need to use the cepstrum script on your input to figure out the needed degradation and arguments for creating the filter.

**Comment by andrixnet** [comment]
Unfortunately not being an expert in the math behind, the only guide I have is http://www.fmwconcepts.com/imagemagick/cepstrum/index.php where the author uses simulated idealized lens/motion blur, thus a very clear circle / well defined diagonal points, but nothing in the way of intepreting less clear cut data. 
Neither does http://www.fmwconcepts.com/imagemagick/fourier_transforms/fourier.html#cepstrum help much. 

For example my cepstrum looks like ... 
![img_2510sq_cep](https://github.com/ImageMagick/ImageMagick/assets/3382777/59d503e8-3efe-4199-82e1-66d382217cc7)

**Reply by fmw42** [reply, marked as answer]
If your cepstrum does not show either two dots or a circle of dots, then your distortion is not uniform motion blur nor uniform lens distortion and cannot be recovered from any of my scripts.  Your distortion is likely a mix of lens distortion and distortion due to non-uniform distance defocus.  Sorry, my scripts won't help in this situation.

**Reply by andrixnet** [reply]
Understood. Thank you for the clarification.
